modules/core/src/main/scala/lucuma/odb/api/repo/TopLevelRepo.scala

```diff
@@ -251,8 +253,97 @@ abstract class TopLevelRepoBase[F[_]: Monad, I: Gid, T: TopLevelModel[I, *]: Eq]
+    }
+
+  protected def shareWithOne[J, M](
+    name: String,
+    id: I,
+    oneId: J,
+    findM: J => State[Tables, ValidatedInput[M]],
+    linkLens: Lens[Tables, OneToMany[J, I]],
+    editedM: M => Long => Event.Edit[M]
+  )(
+    update: (OneToMany[J, I], (J, I)) => OneToMany[J, I]
+  ): F[T] =
+    share[Id, J, M](name, id, oneId, findM, editedM) { vtm =>
+      vtm.traverse_ { _ => linkLens.mod_(links => update(links, (oneId, id))) }.map(_.validNec[InputError].void)
     }
 
+  protected def shareOneWithMany[J, M](
+    name:     String,
+    input:    Sharing[I, J],
+    findM:    J => State[Tables, ValidatedInput[M]],
+    linkLens: Lens[Tables, OneToMany[I, J]],
+    editedM:  M => Long => Event.Edit[M]
+  )(
+    update:   (OneToMany[I, J], IterableOnce[(I, J)]) => OneToMany[I, J]
+  ): F[T] =
+    share(name, input.one, input.many, findM, editedM) { vtm =>
+      vtm.traverse_ { _ => linkLens.mod_(links => update(links, input.tupleLeft)) }.map(_.validNec[InputError].void)
+    }
+
+  protected def shareWithOneUnique[J, M](
+    name: String,
+    id: I,
+    oneId: J,
+    findM: J => State[Tables,ValidatedInput[M]],
+    linkLens: Lens[Tables, OneToManyUnique[J, I]],
+    editedM: M => Long => Event.Edit[M]
+  ): F[T] = 
+    share[Id, J, M](name, id, oneId, findM, editedM){ vtm =>
+      vtm.traverse { _ => 
+        eitherUpdate[Id, OneToManyUnique, Tables, J, I, InputError]((oneId, id), linkLens)(_ + _)}
+          .map(_.andThen(identity))
+    }
+  
+  protected def unshareWithOneUnique[J, M](
+    name: String,
+    id: I,
+    oneId: J,
+    findM: J => State[Tables,ValidatedInput[M]],
+    linkLens: Lens[Tables, OneToManyUnique[J, I]],
+    editedM: M => Long => Event.Edit[M]
+  ): F[T] = 
+    share[Id, J, M](name, id, oneId, findM, editedM) { vtm =>
+      vtm.traverse_ { _ => linkLens.mod_(links => links - ((oneId, id))) }.map(_.validNec[InputError].void)
+    }
+
+  protected def shareOneWithManyUnique[J, M](
+    name: String,
+    input: Sharing[I, J],
+    findM: J => State[Tables, ValidatedInput[M]],
+    linkLens: Lens[Tables, OneToManyUnique[I, J]],
+    editedM: M => Long => Event.Edit[M]
+  ): F[T] = 
+    share(name, input.one, input.many, findM, editedM){ vtm =>
+      vtm.traverse { _ => eitherUpdate(input.tupleLeft, linkLens)(_ ++ _)}.map(_.andThen(identity))
+    }
+
+  protected def unshareOneWithManyUnique[J, M](
+    name: String,
+    input: Sharing[I, J],
+    findM: J => State[Tables, ValidatedInput[M]],
+    linkLens: Lens[Tables, OneToManyUnique[I, J]],
+    editedM: M => Long => Event.Edit[M]
+  ): F[T] = 
+    share(name, input.one, input.many, findM, editedM) { vtm =>
+      vtm.traverse_ { _ => linkLens.mod_(links => links -- input.tupleLeft) }.map(_.validNec[InputError].void)
+    }
+
+  private def eitherUpdate[G[_] : Traverse, H[_, _], S, A, B, E](
+    ab: G[(A, B)], lens: Lens[S, H[A, B]]
+  )(
```

Comment:
I tried to make this function fairly generic, but that turned out to be quite difficult to do. I guess that's why I couldn't find anything "built in" to monocle.

Since it is confined to this file, I could simplify the type parameters by fixing `S -> Table` and `E -> InputError`. With it's current usage, I could even fix `A -> J`, `B -> I` and `H -> OneToManyUnique`. Unless you think it might be useful for anything else.